source/darwin/nw_socket.c

```diff
@@ -0,0 +1,2232 @@
+        args->state = state;
+        args->nw_connection = nw_connection;
+
+        s_socket_acquire_internal_ref(nw_socket);
+
+        aws_task_init(&args->task, s_process_connection_state_changed_task, args, "ConnectionStateChangedTask");
+
+        aws_event_loop_schedule_task_now(nw_socket->event_loop, &args->task);
+
+    } else if (state == nw_connection_state_cancelled) {
+        s_socket_release_internal_ref(nw_socket);
+    }
+
+    s_unlock_base_socket(nw_socket);
+}
+
+static void s_process_listener_success_task(struct aws_task *task, void *args, enum aws_task_status status) {
+    (void)status;
+    (void)task;
+    struct nw_listener_connection_args *task_args = args;
+    struct aws_allocator *allocator = task_args->allocator;
+    struct nw_socket *listener_nw_socket = task_args->nw_socket;
+    int error = task_args->error_code;
+
+    AWS_FATAL_ASSERT(listener_nw_socket && listener_nw_socket->mode == NWSM_LISTENER);
+
+    AWS_LOGF_TRACE(
+        AWS_LS_IO_SOCKET,
+        "id=%p handle=%p: start to process incoming connection.",
+        (void *)listener_nw_socket,
+        (void *)listener_nw_socket->os_handle.nw_listener);
+
+    if (status == AWS_TASK_STATUS_RUN_READY) {
+        s_lock_base_socket(listener_nw_socket);
+        struct aws_socket *listener = listener_nw_socket->base_socket_synced_data.base_socket;
+        AWS_FATAL_ASSERT(listener->accept_result_fn);
+        struct aws_socket *new_socket = NULL;
+
+        if (error) {
+            goto incoming_listener_error_cleanup;
+        }
+
+        new_socket = aws_mem_calloc(allocator, 1, sizeof(struct aws_socket));
+        struct aws_socket_options options = listener->options;
+        error = aws_socket_init(new_socket, allocator, &options);
+        if (error) {
+            goto incoming_listener_error_cleanup;
+        }
+
+        nw_endpoint_t endpoint = nw_connection_copy_endpoint(task_args->new_connection);
+        const char *hostname = nw_endpoint_get_hostname(endpoint);
+        uint16_t port = nw_endpoint_get_port(endpoint);
+
+        if (hostname != NULL) {
+            size_t address_strlen;
+            if (aws_secure_strlen(hostname, AWS_ADDRESS_MAX_LEN, &address_strlen)) {
+                nw_release(endpoint);
+                goto incoming_listener_error_cleanup;
+            }
+            memcpy(new_socket->remote_endpoint.address, hostname, address_strlen);
+            new_socket->remote_endpoint.port = port;
+        }
+        nw_release(endpoint);
+
+        new_socket->io_handle.data.handle = task_args->new_connection;
+        new_socket->io_handle.set_queue = s_client_set_dispatch_queue;
+
+        struct nw_socket *new_nw_socket = new_socket->impl;
+        new_nw_socket->os_handle.nw_connection = task_args->new_connection;
+        new_nw_socket->connection_setup = true;
+
+        // Setup socket state to start read/write operations.
+        s_set_socket_state(new_nw_socket, new_socket, CONNECTED_READ | CONNECTED_WRITE);
```

Comment:
Unsure if it's necessary but we are not holding the synced data lock at the point this is called.

Reply:
Looking into it, We didn't lock here as we are in initializing process, no other process will touch the socket state. Updated the comment to explain it.